modules/core/src/main/java/org/apache/ignite/internal/processors/service/GridServiceProcessor.java

```diff
@@ -1013,19 +829,16 @@ private CancelResult removeServiceFromCache(String name) throws IgniteCheckedExc
      */
     @SuppressWarnings("unchecked")
     public <T> T service(String name) {
-        ctx.security().authorize(name, SecurityPermission.SERVICE_INVOKE, null);
-
-        Collection<ServiceContextImpl> ctxs;
+        if (!enterBusy())
+            return null;
 
-        synchronized (locSvcs) {
-            ctxs = locSvcs.get(name);
-        }
+        try {
+            ctx.security().authorize(name, SecurityPermission.SERVICE_INVOKE, null);
 
-        if (ctxs == null)
-            return null;
+            Collection<ServiceContextImpl> ctxs = Optional.ofNullable(lookupDeployedServiceId(name))
```

Comment:
This two-liner is evil. I need to spend more time understanding it rather than if it were a long code block with `if` statements.

Reply:
Fixed.